File: packages/spectra-py/examples/openrouter_test.py
#!/usr/bin/env python3
"""Spectra Python SDK - Live OpenRouter Test"""

import asyncio
import os
import sys

sys.path.insert(0, ".")

from spectra import Agent, Model, openai


async def main():
    api_key = os.environ.get("OPENROUTER_API_KEY")
    if not api_key:
        print("Error: OPENROUTER_API_KEY not set")
        return

    # Use OpenRouter with base_url via config (if supported)
    # For now, set the key as environment variable
    os.environ["OPENAI_API_KEY"] = api_key

    model = openai("google/gemma-4-26b-a4b-it:free")

    config = {
        "model": {
            "provider": "openrouter",
            "id": model.id,
            "max_tokens": 4096,
        },
        "system_prompt": "You are a helpful coding assistant.",
        "tools": [],
    }

    print(f"=== Testing with {model.id} ===\n")
    agent = Agent(config)
    print("Agent created with native binding!")

    print(
        "\n--- Running: 'Write a Rust function that checks if a number is prime' ---\n"
    )

    async for event in agent.prompt(
        "Write a Rust function that checks if a number is prime. Keep it under 10 lines."
    ):
        if event.get("type") == "message_update":
            print(event.get("delta", ""), end="", flush=True)
        elif event.get("type") == "error":
            print(f"\nError: {event.get('message')}")
        elif event.get("type") == "message_end":
            print(f"\n--- Stop reason: {event.get('stop_reason')} ---")

    print("\n=== Test complete ===")


if __name__ == "__main__":
    asyncio.run(main())

File: crates/spectra-http/examples/openrouter_test.rs
use spectra_http::OpenAIClient;
use spectra_rs::llm::{LlmClient, LlmRequest, Model};
use futures_util::StreamExt;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let api_key = std::env::var("OPENROUTER_API_KEY").expect("OPENROUTER_API_KEY not set");
    
    let client = OpenAIClient::with_api_key(api_key)
        .unwrap()
        .with_base_url("https://openrouter.ai/api/v1/chat/completions");
    
    let model = Model::openai("google/gemma-4-31b-it:free");
    
        let request = LlmRequest {
            model,
            system_prompt: Some("You are a helpful coding assistant.".into()),
            messages: vec![spectra_rs::messages::Message::User(
                spectra_rs::messages::UserMessage::text("Write a Rust function that checks if a number is prime. Keep it under 10 lines.")
            )],
            tools: vec![],
        };

    println!("Calling OpenRouter...");
    let mut stream = client.stream(request).await?;
    println!("\nStreaming response:\n");
    while let Some(event) = stream.next().await {
        match event {
            Ok(spectra_rs::llm::LlmStreamEvent::ContentDelta { delta: spectra_rs::event::ContentDelta::Text { delta: text } }) => {
                print!("{text}");
            }
            Ok(spectra_rs::llm::LlmStreamEvent::ContentDelta { .. }) => {}
            Ok(spectra_rs::llm::LlmStreamEvent::Done { message }) => {
                println!("\n\n=== Done! ===");
                println!("Content: {:?}", message.content);
            }
            Err(e) => {
                eprintln!("Error: {:?}", e);
            }
            _ => {}
        }
    }
    Ok(())
}
